### PR TITLE
More R2R in C

### PR DESCRIPTION
**Pull request by thestr4ng3r**

### CODE REVIEW

**Comment by radare** [comment]
c:\projects\radare2\binr\r2r\run.c(348): error C2065: 'subprocs_mutex': undeclared identifier [C:\projects\radare2\build\binr\r2r\3782362@@r2r@exe.vcxproj]

File: binr/r2r/r2r.c [hunk 1]
```diff
@@ -225,25 +225,42 @@ static void print_diff(const char *actual, const char *expected) {
...
+static R2RProcessOutput *print_runner(const char *file, const char *args[], size_t args_size,
+		const char *envvars[], const char *envvals[], size_t env_size) {
+	size_t i;
+	for (i = 0; i < env_size; i++) {
+		printf ("%s=%s ", envvars[i], envvals[i]);
+	}
+	printf ("%s", file);
+	for (i = 0; i < args_size; i++) {
+		const char *str = args[i];
+		if (strchr (str, '\n') || strchr (str, ' ') || strchr (str, '\'') || strchr (str, '"')) {
```
Comment (radare) [inline binr/r2r/r2r.c]: wtf wat

File: binr/r2r/r2r.c [hunk 2]
```diff
@@ -225,25 +225,42 @@ static void print_diff(const char *actual, const char *expected) {
...
+		if (expect && strcmp (result->proc_out->out, expect) != 0) {
 			printf ("-- stdout\n");
 			print_diff (result->proc_out->out, expect);
 		}
 		expect = result->test->cmd_test->expect_err.value;
-		if (!expect) {
-			break;
-		}
-		if (strcmp (result->proc_out->err, expect) != 0) {
+		if (expect && strcmp (result->proc_out->err, expect) != 0) {
```
Comment (radare) [inline binr/r2r/r2r.c]: unnecessary != 0

File: binr/r2r/r2r.c [hunk 3]
```diff
@@ -225,25 +225,42 @@ static void print_diff(const char *actual, const char *expected) {
...
+static void print_result_diff(R2RRunConfig *config, R2RTestResultInfo *result) {
 	switch (result->test->type) {
 	case R2R_TEST_TYPE_CMD: {
+		r2r_run_cmd_test (config, result->test->cmd_test, print_runner);
 		const char *expect = result->test->cmd_test->expect.value;
-		if (!expect) {
-			expect = "";
-		}
-		if (strcmp (result->proc_out->out, expect) != 0) {
+		if (expect && strcmp (result->proc_out->out, expect) != 0) {
```
Comment (radare) [inline binr/r2r/r2r.c]: drop this != 0

File: binr/r2r/run.c [hunk 4]
```diff
@@ -39,48 +39,59 @@ struct r2r_subprocess_t {
...
-
-		void **it;
-		R2RSubprocess *proc = NULL;
-		r_pvector_foreach (&subprocs, it) {
-			R2RSubprocess *p = *it;
-			if (p->pid == pid) {
-				proc = p;
-				break;
+		ut8 b;
+		ssize_t rd = read (sigchld_pipe[0], &b, 1);
```
Comment (radare) [inline binr/r2r/run.c]: this is not portable

File: binr/r2r/run.c [hunk 5]
```diff
@@ -89,13 +100,33 @@ R_API bool r2r_subprocess_init() {
 	if (!subprocs_mutex) {
 		return false;
 	}
+	if (pipe (sigchld_pipe) == -1) {
+		perror ("pipe");
+		r_th_lock_free (subprocs_mutex);
+		return false;
+	}
+	sigchld_thread = r_th_new (sigchld_th, NULL, 0);
```
Comment (radare) [inline binr/r2r/run.c]: all this stuff should be reimplemented as portable apis in r_util

File: binr/r2r/run.c (hunk 4 above)
Comment (thestr4ng3r) [reply binr/r2r/run.c]: This part of the code is for unix only, so should be fine.

File: binr/r2r/run.c (hunk 5 above)
Comment (thestr4ng3r) [reply binr/r2r/run.c]: Yes, maybe we can move them at some point. Right now it's more convenient to develop them as part of r2r.

**Comment by thestr4ng3r** [comment]
> c:\projects\radare2\binr\r2r\run.c(348): error C2065: 'subprocs_mutex': undeclared identifier [C:\projects\radare2\build\binr\r2r\3782362@@[r2r@exe.vcxproj](mailto:r2r@exe.vcxproj)]

The fact that we still support windows should be considered the bug.

**Pull request merged**